README.md
# waltid-ssikit-vclib
Typesafe implementation of W3C Verifiable Credentials in order to facilitate interoperability among various applications.


### Setup

Add the dependency using Gradle:

    implementation("id.walt:waltid-ssikit-vclib:1.24.0")
    
or Maven:

    <dependency>
        <groupId>id.walt</groupId>
        <artifactId>waltid-ssikit-vclib</artifactId>
        <version>1.24.0</version>
    </dependency>
    
### Create a credential
```kotlin
val verifiableAuthorization: VerifiableCredential = VerifiableAuthorization(
    id = "did:ebsi-eth:00000001/credentials/1872",
    issuer = "did:ebsi:000001234",
    issuanceDate = "2020-08-24T14:13:44Z",
    credentialSubject = VerifiableAuthorization.CredentialSubject1(
        "did:ebsi:00000004321",
        VerifiableAuthorization.CredentialSubject1.NaturalPerson("did:example:00001111")
    ),
    proof = Proof(
        "EcdsaSecp256k1Signature2019",
        "2020-08-24T14:13:44Z",
        "assertionMethod",
        "did:ebsi-eth:000001234#key-1",
        "eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19."
    )
)
```

### Encode a credential to JSON
```kotlin
val json: String = verifiableAuthorization.encode()
// {"@context" : ["https://www.w3.org/2018/credentials/v1", "https://ess ...
```

### Decode a JSON credential
```kotlin
val unknownJson: String = fromInput()

val credential: VerifiableCredential = unknownJson.toCredential()

val issuer = when (credential) {
    is PermanentResidentCard -> credential.issuer
    is VerifiablePresentation -> {
        val vcsInVP = credential.verifiableCredential
        val permanentResidentCardInVP = vcsInVP.any { it is PermanentResidentCard }

        if (permanentResidentCardInVP) {
            (vcsInVP.first { it is PermanentResidentCard } as PermanentResidentCard).issuer
        } else throw IllegalArgumentException("This VP does not contain a PermanentResidentCard")
    }
    else -> throw IllegalArgumentException("No vc of type PermanentResidentCard was found!")
}

// extendable to e.g. take the users address from the PermanentResidentCard, or an Europass (if supplied), or a VerifiableUtilityBill etc...
```

## Adding VC Templates

1. Visit https://vc-creator.walt.id to generate a VC template.
2. Add it to package: id.walt.vclib.credentials.*, compare it to the other templates and complete the implementation.
3. Add the new class to id.walt.vclib.registry.VcTypeRegistry.
4. Run all test-cases, which will generate the JsonSchema and the serialized version of the credential template.
5. Open a Pull-Request on GitHub

## License

The VcLib by walt.id is Open Source software released under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.html).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [1.24.3] - 2022-11-08

## [1.24.2] - 2022-11-08

## [1.24.1] - 2022-11-07

## [1.24.0] - 2022-10-14

-   Updated OpenBadgeCredential due to new JFF requirements

## [1.23.5] - 2022-09-26

## [1.23.5] - 2022-09-26

-   Added velocity email credential

## [1.23.4] - 2022-09-22

-   Updated LegalPerson template to v2206

## [1.23.3] - 2022-09-22

-   Updated LegalPerson template

## [1.23.2] - 2022-09-22

-   Remove invalid contexts from Gaia-X credentials
-   Update Gradle version
-   Change logging type
-   Update schema metadata about Gaia-X credentials

## [1.23.0] - 2022-09-22

-   Added new Gaia-X LegalPerson

-   Added new Gaia-X ParticipantCredential

-   Added new Gaia-X ServiceOfferingCredential

-   Removed legacy LegalPerson

-   Update VCConversionTest to new JWT library version

-   Updated LegalPerson credential type

-   Added explicit blacklisting feature to SchemaValidationTest

-   Updated LegalPerson testing schemas

-   Updated ServiceOffering testing schemas

-   Registered new Gaia-X credential types in VcTypeRegistry

-   Updated vcJsonFromJwt / fromString functions from VerifiableCredential to correspond to work with new JWT & JSON libraries

-   Add debug logging in test mode

-   Fix EuropeanBankIdentity schema URL

-   Update serialized bank identity credential

-   Updated nimbus-jose-jwt from 9.22 to 9.24.4

-   Updated kotlinx-serialization-json from 1.3.2 to 1.4.0

-   Updated jsonschema-generator from 4.24.2 to 4.26.0

-   Updated json-schema-validator from 1.0.69 to 1.0.72

-   Updated json-kotlin-schema from 0.34 to 0.36

-   Updated Kotlin and serialization plugin versions from 1.6.21 to 1.7.10

-   Updated slf4j-simple from 1.7.36 to 2.0.0

-   Updated kotest from 5.3.0 to 5.4.2

-   Updated kotlin-reflect and kotlin-stdlib from 1.6.21 to 1.7.10

## [1.22.0] - 2022-06-20

-   Added Gaia-X LegalPerson Credential 
-   Updated ParticipantCredential and added "hash"

## [1.21.0] - 2022-06-15

-   Instructions how to add VC templates including link to <https://vc-creator.walt.id/>

## [1.20.0] - 2022-06-13

-   Added Open Badge 3.0 credential <https://github.com/walt-id/waltid-ssikit-vclib/pull/51>

## [1.19.0] - 2022-05-19

-   Added Verifiable Mandate <https://github.com/walt-id/waltid-ssikit-vclib/pull/49>

## [1.18.1] - 2022-05-08

-   Bump dependencies 

## [1.18.0] - 2022-03-24

-   Added AmletCredential <https://github.com/walt-id/waltid-ssikit-vclib/pull/46>
-   Added PeerReview credential <https://github.com/walt-id/waltid-ssikit-vclib/pull/47>

## [1.8.0] - 2022-03-22

## [1.17.0] - 2022-03-17

-   Updated Gaia-X Participant Credential according to <https://github.com/walt-id/waltid-ssikit-vclib/issues/43>

## [1.16.0] - 2022-03-12

-   Added missing fields to Europass Credential <https://github.com/walt-id/waltid-ssikit-vclib/pull/41>

## [1.15.0] - 2022-02-16

-   Add missing fields to europass learning specification <https://github.com/walt-id/waltid-ssikit-vclib/pull/40>

## [1.4.1] - 2022-02-03

## [1.14.0] - 2022-02-03

-   Europass update <https://github.com/walt-id/waltid-ssikit-vclib/pull/35> thanks to @ohylli
-   Support of recursive references in JSON schema validation by @atuffreau-bcd <https://github.com/walt-id/waltid-ssikit-vclib/pull/36>

## [1.13.0] - 2022-01-31

## [1.12.0] - 2022-01-27

-   <https://github.com/walt-id/waltid-ssikit-vclib/pull/33>
-   Updated Eurpass and ParticipantCredential

## [1.11.1] - 2022-01-24

-   Remove dummy properties from gaiax kyc credential

## [1.11.0] - 2022-01-24

-   Use github links for credential schema ids, where the actual schema id is not yet known

## [1.10.0] - 2022-01-20

-   Renamed EuropeanBankCredential to EuropeanBankIdentity
-   Pretty VC encoding by: encodePretty()

## [1.9.0] - 2022-01-19

-   Added EuropeanBankCredential

## [1.8.0] - 2022-01-13

-   Moved Gaiax credentails to a dedicated folder
-   Following Gaiax credentials were added (data-models not finalized)
    -   DataConsortium
    -   DataSelfDescription
    -   Iso27001Credential
    -   KybCredential
    -   KybMonoCredential
    -   KycCredential
    -   ParticipantCredential

## [1.7.1] - 2022-01-05

-   fix VC challenge property

## [1.7.0] - 2022-01-03

-   Refactoring of VerifiableCredential data model
-   removal of redundant helper and utility objects (VcLibManager, VcUtils, Helpers and Defaults)

## [1.6.3] - 2021-12-28

-   Added ProofOfResidence credential

## [1.6.2] - 2021-12-03

-   Fixed Json-schema verification for VerifiablePresentations
-   Added VerifiableVaccinationCertificate 
-   Addded GaiaxServiceOffering

## [1.6.1] - 2021-11-30

-   Added GaiaxSelfDescription credential

## [1.6.0] - 2021-11-24

-   Introduced 'Required' annotation for schema generation
-   Added FORBIDDEN_ADDITIONAL_PROPERTIES_BY_DEFAULT for all generated json schemas
-   Added schema-validation function to SchemaService
-   Updated template of VerifiableDiploma & University Degree
-   Added VcUtils (previously in SSI Kit)

## [1.5.4] - 2021-11-20

-   Added eidasLegalIdentifier to VerifiableDiploma

## [1.5.3] - 2021-11-07

-   Added DNSpublicKey to GaiaxCredential

## [1.5.2] - 2021-11-06

-   Removed proofs from templates

## [1.5.1] - 2021-11-06

-   Added GaiaxCredential
-   Added credentialStatus (CredentialStatusList2020) to GaiaxCredential

## [1.5.0] - 2021-11-02

-   Fix/ebsi jwt to credential
-   Fix/get verifiable credential string

## [1.4.9] - 2021-09-25

-   As issuanceDate and expirationDate, validFrom may now have also a non serialized null value according EBSI context.

## [1.4.8] - 2021-09-24

-   Updated VP data model according to EBSI spec
-   expose optional id field on VerifiableCredential abstract interface

## [1.4.7] - 2021-09-13

-   Fix JWT claim named "vp" for verifiable presentation according to EU spec

## [1.4.6] - 2021-09-09

-   JWT-aware credential parsing

src/main/kotlin/id/walt/vclib/credentials/GaiaxSD.kt
package id.walt.vclib.credentials

import com.beust.klaxon.Json
import id.walt.vclib.model.Proof
import id.walt.vclib.model.VerifiableCredential
import id.walt.vclib.registry.VerifiableCredentialMetadata
import id.walt.vclib.schema.SchemaService.PropertyName
import id.walt.vclib.schema.SchemaService.Required

data class GaiaxSD(
    @Json(name = "@context") @field:PropertyName(name = "@context") @field:Required
    var context: List<String> = listOf("https://www.w3.org/2018/credentials/v1"),
    override var id: String?,
    var issuer: String,
    @Json(serializeNull = false) var issuanceDate: String? = null,
    @Json(serializeNull = false) var validFrom: String? = null,
    @Json(serializeNull = false) var expirationDate: String? = null,
    @Json(serializeNull = false) var credentialSubject: CustomCredentialSubject,
    @Json(serializeNull = false) var proof: Proof? = null,
) : VerifiableCredential(type) {
    data class CustomCredentialSubject(
        var id: String, // "Pilot004AIService"
        var type: String, // "Service"
        var hasName: String, // "AIS"
        var description: String, // "AIS demonstrates machine learning application use case."
        var hasVersion: String, // "0.1.0"
        var providedBy: String, // "GAIA-X"
        var hasMarketingImage: String, // https:...
        @Json(serializeNull = false) var hasCertifications: List<String>? = null,
        @Json(serializeNull = false) var utilizes: List<String>? = null,
        @Json(serializeNull = false) var dependsOn: List<String>? = null,
    )

    companion object : VerifiableCredentialMetadata(
        type = listOf("VerifiableCredential", "GaiaxSD"),
        template = {
            GaiaxSD(
                id = "did:ebsi-eth:00000001/credentials/1872",
                issuer = "did:example:456",
                issuanceDate = "2020-08-24T14:13:44Z",
                credentialSubject = CustomCredentialSubject(
                    id = "Pilot004AIService",
                    type = "Service",
                    hasName = "AIS",
                    description = "AIS demonstrates machine learning application use case.",
                    hasVersion = "0.1.0",
                    providedBy = "GAIA-X",
                    hasMarketingImage = "https://www.data-infrastructure.eu/GAIAX/Redaktion/EN/Bilder/UseCases/ai-marketplace-for-product-development.jpg?__blob=normal",
                    hasCertifications = listOf("ISO_27001", "GDPR_Compliance"),
                    utilizes = listOf("ExampleKubernetesService"),
                    dependsOn = listOf("Pilot004DataService"),
                )
            )
        }
    )
}